1762: Add agency information to region

f1sh1918 opened the pull request:
### Short description

Here we add required agency information from freinet to our regions. We need the `agencyId` and `accessKey` to push applicant and card data to freinet in a separate issue

### Proposed changes

<!-- Describe this PR in more detail. -->

- add new database table with freinet agency information
- add FreinetAgencyLoader that fetches agency information on backend startup
- store the agency information in the database and map it to a region
- add a query to get agency information by regionId
- add api tests

### Note
Here is a separate pr to provide backend config and the freinet api access key on the server
https://git.tuerantuer.org/DF/salt/pulls/309/files
Frontend implementation will be done in #1899 

### Side effects

<!-- List all related components that have not been explicitly changed but may be affected by this PR -->

- none

### Testing

1. First you have to add the freinet information & api access key to your `config.local.yml`
You can find the accessKey here: https://passbolt.tuerantuer.org/app/passwords/view/0b693665-7176-4524-a6dd-206fad7c4d8d
2. Run migration
3. Run backend start
4. Agency data should be stored in the database

### Resolved issues

<!-- List all issues which should be closed when this PR is merged. -->

Fixes: #1762

steffenkleinle commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/freinet/util/FreinetAgenciesLoader.kt:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    private fun transformAndFilterAgencyData(agencies: XMLAgencies): List<FreinetApiAgency> {
+        return agencies.agencies.filter { it.agencyId != null && it.ars != null && it.accessKey != null && it.agencyName !== null }.map {
```
:upside_down_face: Could shorten this with [mapNotNull](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/map-not-null.html)

steffenkleinle commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/freinet/webservice/FreinetAgencyQueryService.kt:
```diff
@@ -0,0 +1,40 @@
+        if (projectEntity.project != EAK_BAYERN_PROJECT) {
+            throw NotEakProjectException()
+        }
+        if (!Authorizer.mayViewFreinetAgencyInformationInRegion(admin, regionId)) {
+            throw ForbiddenException()
+        }
```
:thinking: Shouldn't we move this to the start of the function to avoid doing anything the user is not allowed to do?

steffenkleinle commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/freinet/webservice/schema/types/FreinetApiAgency.kt:
```diff
@@ -0,0 +1,8 @@
+package app.ehrenamtskarte.backend.freinet.webservice.schema.types
+
+data class FreinetApiAgency(
+    val arsList: List<String>,
```
:wrench: Can we rename this to something meaningful? What does `ars` stand for?

f1sh1918 replied:
AmtlicherRegionalSchlüssel

f1sh1918 commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/freinet/database/Schema.kt:
```diff
@@ -0,0 +1,25 @@
+    val regionId = reference("regionId", Regions)
+    val agencyId = integer("agencyId").uniqueIndex()
+    val agencyName = varchar("agencyName", 255)
+    val apiAccessKey = varchar("apiAccessKey", 10)
```
good point. i will increase it. what is a good length?

f1sh1918 commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/freinet/database/Setup.kt:
```diff
@@ -0,0 +1,25 @@
+    } else {
+        dbAgency.agencyId = agency.agencyId
+        dbAgency.agencyName = agency.agencyName
+        dbAgency.apiAccessKey = agency.apiAccessKey
```
Not this is a flag that will be manually set by the user and is `false` per default ` #1899

steffenkleinle replied:
Okay, thanks for explaining.

f1sh1918 commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/regions/database/Setup.kt:
```diff
@@ -47,19 +51,25 @@ fun insertOrUpdateRegions() {
+                regionEntity = dbRegion
                 dbRegion.name = eakRegion[1]
                 dbRegion.prefix = eakRegion[0]
                 dbRegion.website = eakRegion[3]
```
hm i do not get why the order matters here and the `dbRegion` has to be moved to the end. Can you explain that to me @steffenkleinle

steffenkleinle replied:
See https://kotlinlang.org/docs/control-flow.html#if-expression (under `Branches of an if expression can be blocks. In this case, the last expression is the value of a block`). Basically, with assignments with an if, the last value in the block is assigned to the variable.

f1sh1918 replied:
thx for the reference!

andrew8er replied:
🔧 Since this is an expression, what about:

```kt
val regionEntity: RegionEntity = dbRegion
    ?.apply {
        name = eakRegion[1]
        prefix = eakRegion[0]
        website = eakRegion[3]
    }
    ?: RegionEntity.new {
        projectId = eakProject.id
        name = eakRegion[1]
        prefix = eakRegion[0]
        regionIdentifier = eakRegion[2]
        website = eakRegion[3]
    }
```

steffenkleinle commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/freinet/webservice/schema/types/XMLAgencies.kt:
```diff
@@ -0,0 +1,12 @@
+data class XMLAgencies(
+    @JsonProperty("mandant")
+    @JacksonXmlElementWrapper(useWrapping = false)
+    var agencies: ArrayList<XMLAgency> = ArrayList()
```
:question: Is there any reason this needs to be `var` instead of `val`?
```suggestion
    var agencies: ArrayList<XMLAgency> = ArrayList()
```

f1sh1918 replied:
no i think not but your suggested changes is wrong :)

steffenkleinle commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/freinet/webservice/schema/types/XMLAgency.kt:
```diff
@@ -0,0 +1,19 @@
+data class XMLAgency(
+    @JsonProperty("agenturid")
+    @JacksonXmlCData
+    var agencyId: String?,
```
Same as above and below.

steffenkleinle commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/freinet/util/FreinetAgenciesLoader.kt:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    private fun transformAndFilterAgencyData(agencies: XMLAgencies): List<FreinetApiAgency> {
+        return agencies.agencies.mapNotNull {
+            FreinetApiAgency(
+                agencyId = it.agencyId!!.toInt(),
+                agencyName = it.agencyName!!,
+                apiAccessKey = it.accessKey!!,
+                officialRegionalKeys = it.officialRegionalKeys?.split(",") ?: emptyList()
+            )
```
:x: Sorry, maybe my previous comment was misunderstandable. This does not really work/will crash if a property is null since you never check whether the properties are null and just use a non-null assertion (`!!`) (even though it seems like nothing is actually null).
`mapNotNull` only filters out `null` after mapping, therefore it does act exactly like `map` in this case. You could do the following (if changing `var` to `val` as mentioned above):
```suggestion
        return agencies.agencies.mapNotNull {
            return@mapNotNull if (it.agencyId != null && it.agencyName != null && it.accessKey != null) {
                FreinetApiAgency(
                    agencyId = it.agencyId.toInt(),
                    agencyName = it.agencyName,
                    apiAccessKey = it.accessKey,
                    officialRegionalKeys = it.officialRegionalKeys?.split(",") ?: emptyList()
                )
            } else null
```

steffenkleinle replied:
Just takes some time to get used to kotlin magic I guess :D

steffenkleinle reviewed (changes requested):
Tested, works as expected. Nice work!

f1sh1918 commented:
> Tested, works as expected. Nice work! The api key is quite short.... 🤔

I know. but thats what we got from freinet. You can tell them your concern about security issues

andrew8er commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/freinet/webservice/FreinetAgencyQueryService.kt:
```diff
@@ -0,0 +1,40 @@
+        if (projectEntity.project != EAK_BAYERN_PROJECT) {
+            throw NotEakProjectException()
+        }
+        val agency = FreinetAgencyRepository.getFreinetAgencyByRegionId(regionId)
+        agency?.let { FreinetAgency(agencyId = it.agencyId, apiAccessKey = agency.apiAccessKey, agencyName = agency.agencyName, dataTransferActivated = agency.dataTransferActivated) }
```
🔧 These two lines can be pulled together using a safe navigation operator, declaring a function level variable is not necessary. I would however, place all arguments of the `FreinetAgency()` ctor on separate lines.

andrew8er commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/freinet/webservice/FreinetAgencyQueryService.kt:
```diff
@@ -0,0 +1,40 @@
+        if (!Authorizer.mayViewFreinetAgencyInformationInRegion(admin, regionId)) {
+            throw ForbiddenException()
+        }
+        val projectEntity = ProjectEntity.find { Projects.project eq project }.firstOrNull()
+            ?: throw ProjectNotFoundException(project)
+
+        if (projectEntity.project != EAK_BAYERN_PROJECT) {
+            throw NotEakProjectException()
+        }
```
🔧 The `if` block below can be pulled up into a `?.let {}` block before the Elvis operator, thus no variable needs to be declared.

f1sh1918 replied:
hm can you give me an example how to do that @andrew8er 
When i put the let after `firstOrNull()` i dunno how i can include both exceptions in the expression

andrew8er replied:
This should be equivalent, if I'm not mistaken:
```kt
        ProjectEntity.find { Projects.project eq project }.firstOrNull()
            ?.let {
                if (it.project != EAK_BAYERN_PROJECT) {
                    throw NotEakProjectException()
                }        
            }
            ?: throw ProjectNotFoundException(project)
```

f1sh1918 replied:
ah alright i thought in can be solved without the using the `if`condition. thx

andrew8er commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/migration/migrations/V0026_AddFreinetAgenciesTable.kt:
```diff
@@ -0,0 +1,27 @@
+        exec(
+            """
+                CREATE TABLE freinetagencies (
+                    id SERIAL PRIMARY KEY ,
```
💭 The recommended way to declare sequence ID columns for modern PostgreSQL is using
[identity columns](https://www.postgresql.org/docs/17/ddl-identity-columns.html).

andrew8er replied:
Yes, an identity column can be (but is not necessarily) a primary key. I don't know if it is a good idea to change the way primary keys are defined mid-project, although I do not see any larger problems with it either. I was just trying to give a hint here.

andrew8er commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/regions/database/Setup.kt:
```diff
@@ -59,6 +63,11 @@ fun insertOrUpdateRegions() {
                 dbRegion.website = eakRegion[3]
+                dbRegion
+            }
+            val agency = agencies.find { agency -> agency.officialRegionalKeys.any { it.startsWith(eakRegion[2]) } }
+            if (agency != null) {
+                insertOrUpdateFreinetRegionInformation(agency, dbFreinetRegionInformation, regionEntity)
```
🔧 The code in the `if` block can be conditionally applied with `?.let {}`. No function level variable necessary here.

andrew8er reviewed (approved):
Some stylistic suggestions.

steffenkleinle commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/auth/service/Authorizer.kt:
```diff
@@ -152,6 +152,11 @@ object Authorizer {
         return user.projectId.value == projectId && user.role == Role.PROJECT_STORE_MANAGER.db_value
     }
 
+    fun mayViewFreinetAgencyInformationInRegion(user: AdministratorEntity, regionId: Int): Boolean {
+        return user.role == Role.REGION_ADMIN.db_value && ProjectEntity.find { Projects.project eq EAK_BAYERN_PROJECT }
+            .single().id.value == user.projectId.value && user.regionId?.value == regionId
+    }
+
```
Good suggestions, I already opened a PR to improve this file though: #1944
I think it would therefore be nice to do further changes there to avoid merge conflicts :)

f1sh1918 replied:
i will keep the current state in my pr

steffenkleinle commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/config/BackendConfiguration.kt:
```diff
@@ -35,7 +36,8 @@ data class ProjectConfig(
     val smtp: SmtpConfig,
-    val matomo: MatomoConfig?
+    val matomo: MatomoConfig?,
+    val freinet: FreinetConfig?
```
Nice, is there a linting/formatter to automate/force these changes?

andrew8er replied:
`ktlint` can do this, as well as the IntelliJ IDEA formatter, but this will generate a lot of changes. Ideally, this would be done, when there are no (or at least very few) PRs contributin to `backend`.

f1sh1918 replied:
I'm not sure if the gradle ktlint plugin can do that but maybe i didn't find the correct package
https://github.com/ankidroid/Anki-Android/pull/14997#issuecomment-1946306075
But maybe we can just provide some additional rules in the `.editorconfig`
Feel free to open a ticket for that.

Currently adding trailing commas will not help because ktlint formating will remove it. So i will keep it as it is

f1sh1918 commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/freinet/database/repos/FreinetAgencyRepository.kt:
```diff
@@ -0,0 +1,10 @@
+
+object FreinetAgencyRepository {
+    fun getFreinetAgencyByRegionId(regionId: Int): FreinetAgenciesEntity? {
+        return FreinetAgenciesEntity.find { FreinetAgencies.regionId eq regionId }.singleOrNull()
```
Alright i guess this is what you mean?
```js
fun getFreinetAgencyByRegionId(regionId: Int): FreinetAgenciesEntity? =
        FreinetAgenciesEntity.find { FreinetAgencies.regionId eq regionId }.singleOrNull()
```

f1sh1918 commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/freinet/util/FreinetAgenciesLoader.kt:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    fun loadAgenciesFromXml(projectConfigs: List<ProjectConfig>): List<FreinetApiAgency> {
+        val bayernConfig =
+            projectConfigs.find { it.id == EAK_BAYERN_PROJECT } ?: throw NotFoundException("Project config not found")
```
hm i think the line break is not really needed here

f1sh1918 commented:
> Some stylistic suggestions.

Thanks for the great review. I think i addressed all comments.
On one or two i was a bit unsure i addressed it correct.
So it would be good to check my last commit

andrew8er commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/migration/migrations/V0026_AddFreinetAgenciesTable.kt:
```diff
@@ -12,7 +12,7 @@ internal class V0026_AddFreinetAgenciesTable : Migration() {
             """
                 CREATE TABLE freinetagencies (
-                    id SERIAL PRIMARY KEY ,
+                     id bigint GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
```
🙃 There's an extra space here.
